plutus-core/cost-model/print-cost-model/Main.hs

```diff
@@ -0,0 +1,235 @@
+     MultipliedSizes    f   -> [ renderLinearFunction f "(x*y)" ]
+     MinSize            f   -> [ renderLinearFunction f "min(x,y)" ]
+     MaxSize            f   -> [ renderLinearFunction f "max(x,y)" ]
+     LinearCost         f   -> [ renderLinearFunction f "x" ]
+     LinearInX          f   -> [ renderLinearFunction f "x" ]
+     LinearInY          f   -> [ renderLinearFunction f "y" ]
+     LinearInZ          f   -> [ renderLinearFunction f "z" ]
+     SubtractedSizes    l c -> [ "if x>y"
+                               , printf "then %s" $ renderLinearFunction l "(x-y)"
+                               , printf "else %d" c
+                               ]
+     ConstAboveDiagonal c m -> [ "if x>y"
+                               , printf "then %s" $ intercalate "\n" (renderModel m)
+                               , printf "else %d" c
+                               ]
+     ConstBelowDiagonal c m -> [ "if x<y"
+                               , printf "then %s" $ intercalate "\n" (renderModel m)
+                               , printf "else %d" c
+                               ]
+     LinearOnDiagonal   f c -> [ "if x==y"
+                               , printf "then %s" $ renderLinearFunction f "x"
+                               , printf "else %d" c
+                               ]
+     -- ^ We're not properly indenting submodels in the above/below diagonal
+     -- cases, but at present our submodels all fit on one line (eg, constant or
+     -- linear).  It seems improbable that we'd ever have a submodel that
+     -- required more than one line because then we'd be dividing the plane up
+     -- into more than two regions.
+
+-- | Take a list of strings and print them line by line, the first with no extra
+-- spaces then the rest preceded by `width` spaces. The assumption is that we'll
+-- already have printed the first part of the first line.
+printListIndented :: Int -> [String] -> IO ()
+printListIndented width l =
+    case l of
+      [] -> pure ()
+      first:rest -> do
+          printf "%s\n" first
+          mapM_ (\s -> printf "%s%s\n" spaces s)  rest
+          where spaces = take width $ repeat ' '
+
+-- | Print a the name of a builtin (the Key below) and then a possibly
+-- multi-line representation of the model, alinged so that each line of the
+-- model has the same indentation.
+printModel :: ModelComponent -> Int -> (Key, CpuAndMemoryModel) -> IO ()
+printModel component width (name, CpuAndMemoryModel cpu mem) = do
+    let model = case component of {Cpu -> cpu; Memory -> mem}
+    printf "%-*s: " width (Key.toString name)
+    printListIndented (width+2) (renderModel model)  -- +2 to account for ": " after builtin name
+
+
+---------------- Command line processing ----------------
+
+usage :: IO a
+usage = do
+  prog <- getProgName
+  printf "Usage: %s [-c|--cpu|-m|--mem|--memory] [-d|--default] [<filename>]\n" prog
+  printf "\n"
+  printf "Print a JSON cost model file in readable form.\n"
```

Comment:
Whenever I see the term 'cost model', i alwyas think 'is this referring to the builtins cost model or the cek machine cost model?'  It turns out that most of the times it refers to the 'builtin cost model', but still it confuses me.

Could we, inside the builtincostmodel.json, merge the cekmachinecosts.json as "special builtins" with 0 arguments and a constant cpu&memory cost?  Then the resulting merged costmodel.json could be used to prettyprint all costs.

Reply:
Good point. I wondered about that myself.  I think we've just been keeping them in separate files because they're produced by different process, but it would be sensible (and I think easy) to merge them into a separate JSON object and put them in one file.  I'll open an issue for that.

My use case for this program was just to look at the builtins, but if we merge the two files I could add an option here to handle that as well (or maybe make get it to work out what sort of object it's been fed).